### PR TITLE
tree-sitter: Update grammars

### PR DESCRIPTION
**Pull request by ghostbuster91**
###### Description of changes

This PR updates the grammars as instructed by

https://github.com/NixOS/nixpkgs/blob/41beb4852c2ab44aceeb7ffc6aa9df61040ef613/pkgs/development/tools/parsing/tree-sitter/default.nix#L28-L32

**Update**

It turned out that `tree-sitter-sql` cannot be built anymore. The reason is that maintainers of that repository decided to move generated parser into a separate branch (see for details:https://github.com/DerekStride/tree-sitter-sql/pull/100). 

This PR allows for specifying an optional parameter `branch` when defining the list of tree-sitter grammars that will be used when fetching grammar parser. 

*Details*: If `branch` key is specified then there is an additional call made to github api to fetch latest commit for the provided branch. It is then passed to `nix-prefech-git` as `rev`. Otherwise the behavior remains the same. I couldn't simply alter `nix-prefetch-git` invocation as the `branch-name` argument is only used there to checkout to a specific branch locally (without tracking the upstream). 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by figsoda** [comment]
Result of `nixpkgs-review pr 232633` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>diffsitter</li>
    <li>tree-sitter-grammars.tree-sitter-sql</li>
  </ul>
</details>
<details>
  <summary>38 packages built:</summary>
  <ul>
    <li>emacsPackages.pygn-mode</li>
    <li>emacsPackages.tree-sitter-ess-r</li>
    <li>emacsPackages.tree-sitter-langs</li>
    <li>tree-sitter-grammars.tree-sitter-c</li>
    <li>tree-sitter-grammars.tree-sitter-clojure</li>
    <li>tree-sitter-grammars.tree-sitter-comment</li>
    <li>tree-sitter-grammars.tree-sitter-cpp</li>
    <li>tree-sitter-grammars.tree-sitter-cuda</li>
    <li>tree-sitter-grammars.tree-sitter-dart</li>
    <li>tree-sitter-grammars.tree-sitter-devicetree</li>
    <li>tree-sitter-grammars.tree-sitter-elisp</li>
    <li>tree-sitter-grammars.tree-sitter-elixir</li>
    <li>tree-sitter-grammars.tree-sitter-elm</li>
    <li>tree-sitter-grammars.tree-sitter-glsl</li>
    <li>tree-sitter-grammars.tree-sitter-haskell</li>
    <li>tree-sitter-grammars.tree-sitter-html</li>
    <li>tree-sitter-grammars.tree-sitter-http</li>
    <li>tree-sitter-grammars.tree-sitter-janet-simple</li>
    <li>tree-sitter-grammars.tree-sitter-json</li>
    <li>tree-sitter-grammars.tree-sitter-ledger</li>
    <li>tree-sitter-grammars.tree-sitter-llvm</li>
    <li>tree-sitter-grammars.tree-sitter-lua</li>
    <li>tree-sitter-grammars.tree-sitter-markdown</li>
    <li>tree-sitter-grammars.tree-sitter-markdown-inline</li>
    <li>tree-sitter-grammars.tree-sitter-nickel</li>
    <li>tree-sitter-grammars.tree-sitter-nix</li>
    <li>tree-sitter-grammars.tree-sitter-perl</li>
    <li>tree-sitter-grammars.tree-sitter-php</li>
    <li>tree-sitter-grammars.tree-sitter-query</li>
    <li>tree-sitter-grammars.tree-sitter-rust</li>
    <li>tree-sitter-grammars.tree-sitter-scheme</li>
    <li>tree-sitter-grammars.tree-sitter-svelte</li>
    <li>tree-sitter-grammars.tree-sitter-tiger</li>
    <li>tree-sitter-grammars.tree-sitter-tlaplus</li>
    <li>tree-sitter-grammars.tree-sitter-tsx</li>
    <li>tree-sitter-grammars.tree-sitter-typescript</li>
    <li>tree-sitter-grammars.tree-sitter-vim</li>
    <li>tree-sitter-grammars.tree-sitter-zig</li>
  </ul>
</details>

**Comment by geri1701** [comment]
Result of `nixpkgs-review pr 232633` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>diffsitter</li>
    <li>tree-sitter-grammars.tree-sitter-sql</li>
  </ul>
</details>
<details>
  <summary>38 packages built:</summary>
  <ul>
    <li>emacsPackages.pygn-mode</li>
    <li>emacsPackages.tree-sitter-ess-r</li>
    <li>emacsPackages.tree-sitter-langs</li>
    <li>tree-sitter-grammars.tree-sitter-c</li>
    <li>tree-sitter-grammars.tree-sitter-clojure</li>
    <li>tree-sitter-grammars.tree-sitter-comment</li>
    <li>tree-sitter-grammars.tree-sitter-cpp</li>
    <li>tree-sitter-grammars.tree-sitter-cuda</li>
    <li>tree-sitter-grammars.tree-sitter-dart</li>
    <li>tree-sitter-grammars.tree-sitter-devicetree</li>
    <li>tree-sitter-grammars.tree-sitter-elisp</li>
    <li>tree-sitter-grammars.tree-sitter-elixir</li>
    <li>tree-sitter-grammars.tree-sitter-elm</li>
    <li>tree-sitter-grammars.tree-sitter-glsl</li>
    <li>tree-sitter-grammars.tree-sitter-haskell</li>
    <li>tree-sitter-grammars.tree-sitter-html</li>
    <li>tree-sitter-grammars.tree-sitter-http</li>
    <li>tree-sitter-grammars.tree-sitter-janet-simple</li>
    <li>tree-sitter-grammars.tree-sitter-json</li>
    <li>tree-sitter-grammars.tree-sitter-ledger</li>
    <li>tree-sitter-grammars.tree-sitter-llvm</li>
    <li>tree-sitter-grammars.tree-sitter-lua</li>
    <li>tree-sitter-grammars.tree-sitter-markdown</li>
    <li>tree-sitter-grammars.tree-sitter-markdown-inline</li>
    <li>tree-sitter-grammars.tree-sitter-nickel</li>
    <li>tree-sitter-grammars.tree-sitter-nix</li>
    <li>tree-sitter-grammars.tree-sitter-perl</li>
    <li>tree-sitter-grammars.tree-sitter-php</li>
    <li>tree-sitter-grammars.tree-sitter-query</li>
    <li>tree-sitter-grammars.tree-sitter-rust</li>
    <li>tree-sitter-grammars.tree-sitter-scheme</li>
    <li>tree-sitter-grammars.tree-sitter-svelte</li>
    <li>tree-sitter-grammars.tree-sitter-tiger</li>
    <li>tree-sitter-grammars.tree-sitter-tlaplus</li>
    <li>tree-sitter-grammars.tree-sitter-tsx</li>
    <li>tree-sitter-grammars.tree-sitter-typescript</li>
    <li>tree-sitter-grammars.tree-sitter-vim</li>
    <li>tree-sitter-grammars.tree-sitter-zig</li>
  </ul>
</details>

**Comment by ghostbuster91** [comment]
Result of `nixpkgs-review pr 232633` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>40 packages built:</summary>
  <ul>
    <li>diffsitter</li>
    <li>emacsPackages.pygn-mode</li>
    <li>emacsPackages.tree-sitter-ess-r</li>
    <li>emacsPackages.tree-sitter-langs</li>
    <li>tree-sitter-grammars.tree-sitter-c</li>
    <li>tree-sitter-grammars.tree-sitter-clojure</li>
    <li>tree-sitter-grammars.tree-sitter-comment</li>
    <li>tree-sitter-grammars.tree-sitter-cpp</li>
    <li>tree-sitter-grammars.tree-sitter-cuda</li>
    <li>tree-sitter-grammars.tree-sitter-dart</li>
    <li>tree-sitter-grammars.tree-sitter-devicetree</li>
    <li>tree-sitter-grammars.tree-sitter-elisp</li>
    <li>tree-sitter-grammars.tree-sitter-elixir</li>
    <li>tree-sitter-grammars.tree-sitter-elm</li>
    <li>tree-sitter-grammars.tree-sitter-glsl</li>
    <li>tree-sitter-grammars.tree-sitter-haskell</li>
    <li>tree-sitter-grammars.tree-sitter-html</li>
    <li>tree-sitter-grammars.tree-sitter-http</li>
    <li>tree-sitter-grammars.tree-sitter-janet-simple</li>
    <li>tree-sitter-grammars.tree-sitter-json</li>
    <li>tree-sitter-grammars.tree-sitter-ledger</li>
    <li>tree-sitter-grammars.tree-sitter-llvm</li>
    <li>tree-sitter-grammars.tree-sitter-lua</li>
    <li>tree-sitter-grammars.tree-sitter-markdown</li>
    <li>tree-sitter-grammars.tree-sitter-markdown-inline</li>
    <li>tree-sitter-grammars.tree-sitter-nickel</li>
    <li>tree-sitter-grammars.tree-sitter-nix</li>
    <li>tree-sitter-grammars.tree-sitter-perl</li>
    <li>tree-sitter-grammars.tree-sitter-php</li>
    <li>tree-sitter-grammars.tree-sitter-query</li>
    <li>tree-sitter-grammars.tree-sitter-rust</li>
    <li>tree-sitter-grammars.tree-sitter-scheme</li>
    <li>tree-sitter-grammars.tree-sitter-sql</li>
    <li>tree-sitter-grammars.tree-sitter-svelte</li>
    <li>tree-sitter-grammars.tree-sitter-tiger</li>
    <li>tree-sitter-grammars.tree-sitter-tlaplus</li>
    <li>tree-sitter-grammars.tree-sitter-tsx</li>
    <li>tree-sitter-grammars.tree-sitter-typescript</li>
    <li>tree-sitter-grammars.tree-sitter-vim</li>
    <li>tree-sitter-grammars.tree-sitter-zig</li>
  </ul>
</details>

**Comment by geri1701** [comment]
Result of `nixpkgs-review pr 232633` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>40 packages built:</summary>
  <ul>
    <li>diffsitter</li>
    <li>emacsPackages.pygn-mode</li>
    <li>emacsPackages.tree-sitter-ess-r</li>
    <li>emacsPackages.tree-sitter-langs</li>
    <li>tree-sitter-grammars.tree-sitter-c</li>
    <li>tree-sitter-grammars.tree-sitter-clojure</li>
    <li>tree-sitter-grammars.tree-sitter-comment</li>
    <li>tree-sitter-grammars.tree-sitter-cpp</li>
    <li>tree-sitter-grammars.tree-sitter-cuda</li>
    <li>tree-sitter-grammars.tree-sitter-dart</li>
    <li>tree-sitter-grammars.tree-sitter-devicetree</li>
    <li>tree-sitter-grammars.tree-sitter-elisp</li>
    <li>tree-sitter-grammars.tree-sitter-elixir</li>
    <li>tree-sitter-grammars.tree-sitter-elm</li>
    <li>tree-sitter-grammars.tree-sitter-glsl</li>
    <li>tree-sitter-grammars.tree-sitter-haskell</li>
    <li>tree-sitter-grammars.tree-sitter-html</li>
    <li>tree-sitter-grammars.tree-sitter-http</li>
    <li>tree-sitter-grammars.tree-sitter-janet-simple</li>
    <li>tree-sitter-grammars.tree-sitter-json</li>
    <li>tree-sitter-grammars.tree-sitter-ledger</li>
    <li>tree-sitter-grammars.tree-sitter-llvm</li>
    <li>tree-sitter-grammars.tree-sitter-lua</li>
    <li>tree-sitter-grammars.tree-sitter-markdown</li>
    <li>tree-sitter-grammars.tree-sitter-markdown-inline</li>
    <li>tree-sitter-grammars.tree-sitter-nickel</li>
    <li>tree-sitter-grammars.tree-sitter-nix</li>
    <li>tree-sitter-grammars.tree-sitter-perl</li>
    <li>tree-sitter-grammars.tree-sitter-php</li>
    <li>tree-sitter-grammars.tree-sitter-query</li>
    <li>tree-sitter-grammars.tree-sitter-rust</li>
    <li>tree-sitter-grammars.tree-sitter-scheme</li>
    <li>tree-sitter-grammars.tree-sitter-sql</li>
    <li>tree-sitter-grammars.tree-sitter-svelte</li>
    <li>tree-sitter-grammars.tree-sitter-tiger</li>
    <li>tree-sitter-grammars.tree-sitter-tlaplus</li>
    <li>tree-sitter-grammars.tree-sitter-tsx</li>
    <li>tree-sitter-grammars.tree-sitter-typescript</li>
    <li>tree-sitter-grammars.tree-sitter-vim</li>
    <li>tree-sitter-grammars.tree-sitter-zig</li>
  </ul>
</details>

**Review by geri1701 — APPROVED** [review]

**Comment by figsoda** [comment]
Instead of specifying a branch, we can also do this:
```patch
--- a/pkgs/development/tools/parsing/tree-sitter/default.nix
+++ b/pkgs/development/tools/parsing/tree-sitter/default.nix
@@ -62,12 +62,14 @@ let
           inherit version;
           src = grammar.src or (fetchGrammar grammar);
           location = grammar.location or null;
+          generate = grammar.generate or false;
         };
       grammars' = import ./grammars { inherit lib; } // extraGrammars;
       grammars = grammars' //
         { tree-sitter-ocaml = grammars'.tree-sitter-ocaml // { location = "ocaml"; }; } //
         { tree-sitter-ocaml-interface = grammars'.tree-sitter-ocaml // { location = "interface"; }; } //
         { tree-sitter-org-nvim = grammars'.tree-sitter-org-nvim // { language = "org"; }; } //
+        { tree-sitter-sql = grammars'.tree-sitter-sql // { generate = true; }; } //
         { tree-sitter-typescript = grammars'.tree-sitter-typescript // { location = "typescript"; }; } //
         { tree-sitter-tsx = grammars'.tree-sitter-typescript // { location = "tsx"; }; } //
         { tree-sitter-markdown = grammars'.tree-sitter-markdown // { location = "tree-sitter-markdown"; }; } //
```

**Comment by ghostbuster91** [comment]
> Instead of specifying a branch, we can also do this:
> 
> ```diff
> --- a/pkgs/development/tools/parsing/tree-sitter/default.nix
> +++ b/pkgs/development/tools/parsing/tree-sitter/default.nix
> @@ -62,12 +62,14 @@ let
>            inherit version;
>            src = grammar.src or (fetchGrammar grammar);
>            location = grammar.location or null;
> +          generate = grammar.generate or false;
>          };
>        grammars' = import ./grammars { inherit lib; } // extraGrammars;
>        grammars = grammars' //
>          { tree-sitter-ocaml = grammars'.tree-sitter-ocaml // { location = "ocaml"; }; } //
>          { tree-sitter-ocaml-interface = grammars'.tree-sitter-ocaml // { location = "interface"; }; } //
>          { tree-sitter-org-nvim = grammars'.tree-sitter-org-nvim // { language = "org"; }; } //
> +        { tree-sitter-sql = grammars'.tree-sitter-sql // { generate = true; }; } //
>          { tree-sitter-typescript = grammars'.tree-sitter-typescript // { location = "typescript"; }; } //
>          { tree-sitter-tsx = grammars'.tree-sitter-typescript // { location = "tsx"; }; } //
>          { tree-sitter-markdown = grammars'.tree-sitter-markdown // { location = "tree-sitter-markdown"; }; } //
> ```

That would be also possible. However, I think that the ability to chose branch might be handy so I would keep it the way how it is, unless you have a strong preference toward `generate`.

**Comment by figsoda** [comment]
It would be easier for me to review since it's simpler and I'm also not familiar with the update script, but either option would work

File: pkgs/development/tools/parsing/tree-sitter/update_impl.py [hunk 1]
```diff
@@ -106,9 +106,23 @@ def fetchRepo() -> None:
...
+                    # otherwise fallback to using HEAD
+                    if "branch" in jsonArg:
+                        branch = jsonArg.get("branch")
+                        branch_out = run_cmd(curl_github_args(
+                            token,
+                            url=f"https://api.github.com/repos/{quote(orga)}/{quote(repo)}/branches/{quote(branch)}"
+                        ))
+                        match curl_result(branch_out):
+                            case {"commit": {"sha": commit_hash}}:
+                                release = commit_hash
```
Comment (figsoda) [inline pkgs/development/tools/parsing/tree-sitter/update_impl.py:119]: I think we can just do `f"refs/heads/{branch}"` here without the extra api request, `nix-prefetch-git` still outputs a sha as `rev` as expected, using the command from `nix_prefetch_git_args`

```console
$ nix-prefetch-git --quiet --no-deepClone --url https://github.com/derekstride/tree-sitter-sql --rev refs/heads/gh-pages
{
  "url": "https://github.com/derekstride/tree-sitter-sql",
  "rev": "721087c8819cda10ca37f974e914ab9be46b290f",
  "date": "2023-05-19T15:56:01+00:00",
  "path": "/nix/store/5mbri1m7rrwmvhzp5nz3bs1pdzj87fni-tree-sitter-sql",
  "sha256": "10njfi84b78z036bkgbrxnxgy9av9yprxq1p2zla87w0g6ixqva7",
  "fetchLFS": false,
  "fetchSubmodules": false,
  "deepClone": false,
  "leaveDotGit": false
}
```

Comment (ghostbuster91) [reply pkgs/development/tools/parsing/tree-sitter/update_impl.py:119]: Right, great :) this will simplify things.

**Review by figsoda — APPROVED** [review]

**Pull request merged**